Implement Data Normalization/Parsing for Trading Economics Calendar Data

gsalaz98 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
<!--- Describe your changes in detail -->
Normalizes Trading Economics calendar data to decimal values.

Note that these changes will most likely cause PR #3588 to have merge conflicts
#### Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Closes #3595 
#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Ensure consistency across custom data types
#### Requires Documentation Change
<!--- Please indicate if these changes will require updates to documentation, and if so, specify what changes are required -->
Yes - data library documentation
#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Ran algorithm locally. Processed data locally from 2001 to 2019
#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

Jay-Jay-D commented on ToolBox/TradingEconomicsDataDownloader/TradingEconomicsCalendarDownloader.cs:
```diff
@@ -190,5 +192,247 @@ public override Task<string> Get(DateTime startUtc, DateTime endUtc)
+        /// The economic calendar covers around 1600 events for more than 150 countries a month.
+        /// https://docs.tradingeconomics.com/#events
+        /// </summary>
+        private class TradingEconomicsCalendarRaw
```
Is there a reason for having this new class here instead of a new file?

gsalaz98 replied:
I figured that it was the only place where we're ever going to use it, since we transform the data to a new format shortly afterwards and never deal with raw data aside from here. Maybe it might be a worthy endeavor to put it in a new file now that you mention it for future maintenance and extensibility 👍

Martin-Molinero replied:
If the only difference between the `Raw` version and the one used by Lean is the strings to decimals, it seems to me it might not be worth adding this extra class? Maybe this `Raw` version could be a derivate of the other one so we avoid some duplication?

Martin-Molinero commented on ToolBox/TradingEconomicsDataDownloader/TradingEconomicsCalendarDownloader.cs:
```diff
@@ -190,5 +192,247 @@ public override Task<string> Get(DateTime startUtc, DateTime endUtc)
+                    Ticker = Ticker,
+                    TESymbol = TESymbol,
+
+                    IsPercentage = Actual.EndsWith("%") || Previous.EndsWith("%") || Forecast.EndsWith("%") || TradingEconomicsForecast.EndsWith("%")
```
The `||` operations to defined `IsPercentage` worries me, how will the user be able to determine which value is a percentage or not? 
It seems these values are all referring to the same 'thing', have we seen a case where some are a `%` and some are not? 
Could we normilize them before hand so that all values are a percentage or not?

Martin-Molinero commented on ToolBox/TradingEconomicsDataDownloader/TradingEconomicsCalendarDownloader.cs:
```diff
@@ -190,5 +192,247 @@ public override Task<string> Get(DateTime startUtc, DateTime endUtc)
+                };
+            }
+
+            private decimal? ParseDecimal(string figure)
```
I think it might be a good idea to move this parser to a shared `utils` location and add unit tests for it

gsalaz98 commented:
Addressed reviews.

One thing that has me left uneasy is determining which value is a percentage or not. Since all of the figures are reported in the same unit, I made the assumption that if any value is a percentage then all of them are.

Martin-Molinero commented on Tests/ToolBox/TradingEconomicsCalendarTests.cs:
```diff
@@ -0,0 +1,41 @@
+        public void DecimalIsParsedCorrectly(string value, double? expected, bool inPercentage)
+        {
+            // Cast inside since we can't pass in decimal values through TestCase attributes
+            Assert.AreEqual((decimal?)expected, TradingEconomicsCalendarDownloader.ParseDecimal(value, inPercentage));
```
Nice 👍  thanks for the tests

Martin-Molinero commented on ToolBox/TradingEconomicsDataDownloader/TradingEconomicsCalendarDownloader.cs:
```diff
@@ -94,13 +97,27 @@ public override bool Run()
+                        rawData["Revised"] = ParseDecimal(rawData["Revised"].Value<string>(), inPercentage);
+                    }
+
+                    var collection = rawCollection.ToObject<List<TradingEconomicsCalendar>>();
```
Agree with @mchandschuh idea! If this were a custom JSON converter it would be beautiful ⭐️

gsalaz98 replied:
I played around with using a JSON Converter, but ultimately it didn't pan out, so this was the next best solution I could think of
